compiler/src/iree/compiler/Codegen/Common/GPU/GPUConvertToCoalescedDMA.cpp

```diff
@@ -33,223 +36,433 @@ namespace mlir::iree_compiler {
 
 namespace {
 
-// Helper function to validate and extract constant tile sizes from
-// OpFoldResults
-static FailureOr<SmallVector<int64_t>>
-getConstantTileSizes(ArrayRef<OpFoldResult> tileSizes) {
-  SmallVector<int64_t> staticTileSizes;
-  for (OpFoldResult ofr : tileSizes) {
-    std::optional<int64_t> tileSize = getConstantIntValue(ofr);
-    if (!tileSize) {
-      return failure();
-    }
-    staticTileSizes.push_back(*tileSize);
+/// Create GPU warp mapping attributes for the given rank.
+static SmallVector<Attribute> getWarpMapping(MLIRContext *ctx, int64_t rank) {
+  SmallVector<Attribute> mapping;
+  for (int64_t i = 0; i < rank; ++i) {
+    auto mappingId = static_cast<gpu::MappingId>(
+        static_cast<int>(gpu::MappingId::LinearDim0) + (rank - 1 - i));
+    mapping.push_back(gpu::GPUWarpMappingAttr::get(ctx, mappingId));
   }
-  return staticTileSizes;
+  return mapping;
 }
 
-// Helper function to create GPU mapping attributes for forall loops
-static SmallVector<Attribute>
-createGPUMappingAttrs(MLIRContext *ctx, int64_t rank, bool useWarpMapping) {
+/// Create GPU thread mapping for lane mapping.
+/// Returns a single-element array with gpu.lane_id<0>.
+static SmallVector<Attribute> getThreadMapping(MLIRContext *ctx,
+                                               int64_t numLoops) {
   SmallVector<Attribute> mapping;
+  // Since we only tile the innermost dimension, we only have one loop
+  // Map it to gpu.lane_id<0>
+  mapping.push_back(IREE::GPU::LaneIdAttr::get(ctx, 0));
+  return mapping;
+}
+
+/// Helper to compute thread number of threads based on translation_info.
+/// Uses the subgroup_size from translation_info for thread-level tiling.
+static SmallVector<OpFoldResult>
+computeThreadNumThreadsImpl(OpBuilder &builder, Operation *op,
+                            RankedTensorType outputType) {
+  // Check that this operation has the use_global_load_dma config.
+  auto dmaConfig = getLoweringConfig<IREE::GPU::UseGlobalLoadDMAAttr>(op);
+  if (!dmaConfig) {
+    return {};
+  }
+
+  // Get the function containing this operation.
+  auto funcOp = op->getParentOfType<FunctionOpInterface>();
+  if (!funcOp) {
+    return {};
+  }
+
+  // Get subgroup size from translation_info.
+  std::optional<int64_t> subgroupSize = getSubgroupSize(funcOp);
+  if (!subgroupSize) {
+    return {};
+  }
+
+  // Only tile over the innermost dimension.
+  // Return a 1-element array with the subgroup size.
+  // SCF tiling will tile the last (innermost) dimension.
+  SmallVector<OpFoldResult> numThreads;
+  numThreads.push_back(builder.getIndexAttr(*subgroupSize));
+  return numThreads;
+}
+
+/// Check if the given forall op has warp mapping.
+static bool hasWarpMapping(scf::ForallOp forallOp) {
+  if (!forallOp) {
+    return false;
+  }
+
+  auto mapping = forallOp.getMapping();
+  if (!mapping.has_value()) {
+    return false;
+  }
+
+  return llvm::any_of(mapping.value(), [](Attribute attr) {
```

Comment:
ubernit: all_of for safety. Realistically verification for the mapping could be improved on scf.forall but w/e.